### PR TITLE
Enable JIT in PCRE to improve regular expressions performance

### PR DESCRIPTION
**Pull request by scriptum**
Regular expressions in Nim are slow because they are not using JIT feature of PCRE. Idea of using JIT in pcre_study based on grep source. Performance impact is **huge**.

There is still bottleneck in `nre`, it is 2-3 times slower than `re`.

### CODE REVIEW

File: lib/impure/nre.nim [hunk 1]
```diff
@@ -432,7 +432,12 @@ proc initRegex(pattern: string, flags: int, study = true): Regex =
 
   if study:
     # XXX investigate JIT
```
Comment (flaviut) [inline lib/impure/nre.nim]: It appears to me as if JIT has been investigated and this line can be safely removed :grin:.

File: lib/impure/re.nim [hunk 2]
```diff
@@ -214,7 +219,7 @@ proc find*(s: string, pattern: Regex, start = 0): int =
   var
     rtarray = initRtArray[cint](3)
     rawMatches = rtarray.getRawData
-    res = pcre.exec(pattern.h, nil, s, len(s).cint, start.cint, 0'i32,
```
Comment (flaviut) [inline lib/impure/re.nim]: So re.nim never used study data in find? Nice catch!

File: lib/impure/nre.nim (hunk 1 above)
Comment (scriptum) [inline lib/impure/nre.nim]: Removed

**Comment by flaviut** [comment]
```
<flaviut> dom96: can I/you merge #4814?
<dom96> flaviut: I would have done it already but it seemed risky given 0.15.2 was meant to be released soon
...
<dom96> I think it's better if we wait. The downside is slower performance, but the potential issues could be huge. I don't want 0.15.2 to be released with a broken 're' module.
```

**Comment by scriptum** [comment]
Ok. There are some issues with nre performance, it uses heap allocation too frequently, but seems that I'm not able fix that.

**Comment by flaviut** [comment]
What are you using nre for?

When designing nre, I was really focused on making the API easy to use, and less focused on performance.

**Comment by scriptum** [comment]
I tried to use regular expressions for parsing. After some tests I decided to rewrite regular expressions in hot loops to plain Nim code because it was faster for my purposes. Next I noticed that grep is extremely fast for parsing even when using regular expressions and found this JIT issue.

When I tried to study cause of bad PCRE performance in Nim, I noticed that nre always slower than re especially while parsing big files. Profiler shows that there are many memory allocations when I use nre. Probably it could be optimized with term rewriting marcros and still be easy to use. E.g. allocate matches sequence while regex object created instead of every exec. Since regex compiled, size of matches sequence shoudn't change. One interesting thing is that regular expressions are almost always literals and could be prepared, analyzed and optimized in compile time, not runtime. That is a thing for what Nim is good.

Recently I used Python for similar tasks and I wondered that Python has its own regex implementation. And it is really, really slow.

But now, with that patch, Pegs module could be discarded (in my code at least) because it isn't easy to use and has weird syntax. I'd prefer to have hyperfast DFA implementation from grep because in 90% of cases regular expressions are pretty simple.

**Comment by flaviut** [comment]
> E.g. allocate matches sequence while regex object created instead of every exec.

It's impossible to do that safely without keeping track of lifetimes. It may be possible to implement object lifetimes, but it's going to be difficult.

---------------

> I'd prefer to have hyperfast DFA implementation from grep because in 90% of cases regular expressions are pretty simple. Grep doesn't use JIT but it is faster than PCRE+JIT and much faster than pegs.

PCRE has pcre_dfa_exec(), I don't know if that's what you're looking for. Adding a DFA flag to nre might be nice though.

-------------

I took a look at `re.nim` to see how it avoids allocating. It looks like it uses `lib/pure/collections/rtarrays.nim` to avoid allocating when (number of matches+1)*3 ≤ 10.

A possible optimization for nre might be to avoid allocating an `ovector` (known as `pcreMatchBounds` in nre) if there are no captures and no back-references. See [man pcreapi](http://www.pcre.org/original/doc/html/pcreapi.html), especially around

> If neither the actual string matched nor any captured substrings are of interest, pcre_exec() may be called with ovector passed as NULL and ovecsize as zero. However, if the pattern contains back references and the ovector is not big enough to remember the related substrings, PCRE has to get additional memory for use during matching. Thus it is usually advisable to supply an ovector of reasonable size.

**Comment by scriptum** [comment]
> It's impossible to do that safely without keeping track of lifetimes. It may be possible to implement object lifetimes, but it's going to be difficult.

Now funny part: I never used group capture in my tests:)
One idea I can suppose: count captures at regex study step, next allocate array if it is small enough on stack otherwise on heap. But probably this is GC issue - many small allocations with same size cause this effect.

> PCRE has pcre_dfa_exec(), I don't know if that's what you're looking for. Adding a DFA flag to nre might be nice though.

I mean not dfa itself but dfa implementation from grep which is the fastest I ever seen (for simple expressions at least). c2nim failed convert it:)

**Comment by Araq** [comment]
> I mean not dfa itself but dfa implementation from grep which is the fastest I ever seen (for simple expressions at least). c2nim failed convert it:)

Please checkout my lexim project if you're interested in this. It's a lexer generator in desparate need of a more serious maintainer.

**Comment by scriptum** [comment]
> Please checkout my lexim project if you're interested in this. It's a lexer generator in desparate need of a more serious maintainer.

@Araq lexim project implements very interesting idea and shows show regular expressions could be optimized with macro. But I cannot run it on full speed due to bug with memfiles (gcc cannot compile generated C code). I tried very simplified regex-dna benchmark:

```Nim
import lexim, memfiles

proc main =
  # var f = memfiles.open("../regex-dna/in.txt")
  var c = 0
  for line in "../regex-dna/in.txt".lines():
  # BUG
  # for line in f.lines():
    var pos = 0
    while pos < line.len:
      match line, pos:
      of r"agggtaaa|tttaccct": inc c
      of r".": discard
  # f.close()
  echo c

main()
```

Without memfiles it works but PCRE with JIT is faster for this test. Probably there is a better way but I found only one example.

Here is the possible bottleneck:

```C
{
	if (!(((NU8)(input0->data[pos0])) >= ((NU8)(1)) && ((NU8)(input0->data[pos0])) <= ((NU8)(96)) || ((NU8)(input0->data[pos0])) >= ((NU8)(98)) && ((NU8)(input0->data[pos0])) <= ((NU8)(115)) || ((NU8)(input0->data[pos0])) >= ((NU8)(117)) && ((NU8)(input0->data[pos0])) <= ((NU8)(255)))) goto LA14;
	pos0 += ((NI) 1);
	goto NIMSTATE_9;
}
```

**Pull request merged**